// @tsed/common/lib/mvc/registries/ControllerRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const di_1 = require("@tsed/di");
const ControllerProvider_1 = require("../class/ControllerProvider");
const ExpressRouter_1 = require("../services/ExpressRouter");
// tslint:disable-next-line: variable-name
exports.ControllerRegistry = di_1.GlobalProviders.createRegistry(di_1.ProviderType.CONTROLLER, ControllerProvider_1.ControllerProvider, {
    injectable: false,
    buildable: true,
    onInvoke(provider, locals, designParamTypes) {
        if (!locals.has(ExpressRouter_1.ExpressRouter)) {
            locals.set(ExpressRouter_1.ExpressRouter, provider.router);
        }
    }
});
/**
 * Add a new controller in the `ProviderRegistry`. This controller will be built when `InjectorService` will be loaded.
 *
 * #### Example
 *
 * ```typescript
 * import {registerController, InjectorService} from "@tsed/common";
 *
 * export default class MyController {
 *     constructor(){}
 *     transform() {
 *         return "test";
 *     }
 * }
 *
 * registerController({provide: MyController});
 * // or
 * registerController(MyController);
 *
 * const injector = new InjectorService();
 * injector.load();
 *
 * const myController = injector.get<MyController>(MyController);
 * myController.getFoo(); // test
 * ```
 *
 * @param provider Provider configuration.
 */
exports.registerController = di_1.GlobalProviders.createRegisterFn(di_1.ProviderType.FILTER);
