// ts-log-debug/lib/logger/utils/tableUtils.js
"use strict";
/**
 * @module logger
 */
/** */
Object.defineProperty(exports, "__esModule", { value: true });
function charRepeater(x, char) {
    if (char === void 0) { char = " "; }
    var res = "";
    while (x--)
        res += char;
    return res;
}
exports.charRepeater = charRepeater;
/**
 *
 * @returns {string}
 */
function buildStartLine(fields, settings) {
    var line = "┌";
    var list = Object.keys(fields);
    list.forEach(function (key, index) {
        if (index !== 0 && index !== list.length) {
            line += "┬";
        }
        line += charRepeater(fields[key] + 2 * settings.padding, "─");
    });
    line += "┐";
    return line;
}
exports.buildStartLine = buildStartLine;
/**
 *
 * @param fields
 * @param settings
 * @returns {string}
 */
function buildEndLine(fields, settings) {
    var line = "└";
    var list = Object.keys(fields);
    list.forEach(function (key, index) {
        if (index !== 0 && index !== list.length) {
            line += "┴";
        }
        line += charRepeater(fields[key] + 2 * settings.padding, "─");
    });
    line += "┘";
    return line;
}
exports.buildEndLine = buildEndLine;
/**
 *
 * @param fields
 * @param settings
 * @param char
 * @returns {string}
 */
function buildLine(fields, settings, char) {
    if (char === void 0) { char = "─"; }
    var line = "";
    Object.keys(fields).forEach(function (key) {
        line += "│";
        line += charRepeater(fields[key] + 2 * settings.padding, char);
    });
    line += "│";
    return line;
}
exports.buildLine = buildLine;
/**
 *
 */
function buildLineData(scope, fields, settings) {
    var line = "";
    Object.keys(fields).forEach(function (key) {
        line += "│ ";
        line += scope[key];
        line += charRepeater(fields[key] + 2 * (settings.padding - 1) - scope[key].length, " ");
        line += " ";
    });
    line += "│";
    return line;
}
exports.buildLineData = buildLineData;
/**
 *
 * @param list
 * @param settings
 * @returns {string}
 */
function drawTable(list, settings) {
    if (settings === void 0) { settings = {}; }
    settings.padding = settings.padding || 1;
    if (settings.header === undefined) {
        settings.header = {};
        Object.keys(list[0]).forEach(function (key) { return settings.header[key] = key; });
    }
    var fields = {};
    // Calculate width for each column
    Object.keys(settings.header).forEach(function (key) { return fields[key] = settings.header[key].length; });
    list.forEach(function (route) {
        Object.keys(fields).forEach(function (key) {
            return fields[key] = Math.max(("" + route[key]).length, fields[key]);
        });
    });
    var output = "";
    output += buildStartLine(fields, settings) + "\n";
    output += buildLineData(settings.header, fields, settings) + "\n";
    list.forEach(function (scope) {
        output += buildLine(fields, settings) + "\n";
        output += buildLineData(scope, fields, settings) + "\n";
    });
    output += buildEndLine(fields, settings);
    return output;
}
exports.drawTable = drawTable;
