// ts-log-debug/lib/layouts/components/JsonLayout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/**
 * @module layouts
 */
/** */
var BaseLayout_1 = require("../class/BaseLayout");
var layout_1 = require("../decorators/layout");
var inspectUtils_1 = require("../utils/inspectUtils");
/**
 *
 * @private
 */
var JsonLayout = /** @class */ (function (_super) {
    tslib_1.__extends(JsonLayout, _super);
    function JsonLayout() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    JsonLayout.prototype.transform = function (loggingEvent, timezoneOffset) {
        var log = {
            startTime: loggingEvent.startTime,
            categoryName: loggingEvent.categoryName,
            level: loggingEvent.level.toString(),
            data: loggingEvent.data,
            context: loggingEvent.context
        };
        log.data = log.data.map(function (data) { return inspectUtils_1.formatLogData([data]); });
        return JSON.stringify(log) + (this.config["separator"] || "");
    };
    JsonLayout = tslib_1.__decorate([
        layout_1.Layout({ name: "json" })
    ], JsonLayout);
    return JsonLayout;
}(BaseLayout_1.BaseLayout));
exports.JsonLayout = JsonLayout;
