// @tsed/di/lib/registries/ProviderRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Provider_1 = require("../class/Provider");
const Providers_1 = require("../class/Providers");
const interfaces_1 = require("../interfaces");
/**
 *
 * @type {Providers}
 */
// tslint:disable-next-line: variable-name
exports.GlobalProviders = new Providers_1.Providers();
/**
 *
 * @type {Providers}
 */
// tslint:disable-next-line: variable-name
exports.ProviderRegistry = exports.GlobalProviders.getRegistry(interfaces_1.ProviderType.PROVIDER);
/**
 *
 * @type {Registry<Provider<any>, IProvider<any>>}
 */
exports.GlobalProviders.createRegistry(interfaces_1.ProviderType.SERVICE, Provider_1.Provider, {
    injectable: true,
    buildable: true
});
/**
 *`
 * @type {Registry<Provider<any>, IProvider<any>>}
 */
exports.GlobalProviders.createRegistry(interfaces_1.ProviderType.FACTORY, Provider_1.Provider, {
    injectable: true,
    buildable: false
});
/**
 * Add a new factory in the `ProviderRegistry`.
 *
 * #### Example with symbol definition
 *
 *
 * ```typescript
 * import {registerFactory, InjectorService} from "@tsed/common";
 *
 * export interface IMyFooFactory {
 *    getFoo(): string;
 * }
 *
 * export type MyFooFactory = IMyFooFactory;
 * export const MyFooFactory = Symbol("MyFooFactory");
 *
 * registerFactory(MyFooFactory, {
 *      getFoo:  () => "test"
 * });
 *
 * // or
 *
 * registerFactory({provide: MyFooFactory, instance: {
 *      getFoo:  () => "test"
 * }});
 *
 * @Service()
 * export class OtherService {
 *      constructor(@Inject(MyFooFactory) myFooFactory: MyFooFactory){
 *          console.log(myFooFactory.getFoo()); /// "test"
 *      }
 * }
 * ```
 *
 * > Note: When you use the factory method with Symbol definition, you must use the `@Inject()`
 * decorator to retrieve your factory in another Service. Advice: By convention all factory class name will be prefixed by
 * `Factory`.
 *
 * #### Example with class
 *
 * ```typescript
 * import {InjectorService, registerFactory} from "@tsed/common";
 *
 * export class MyFooService {
 *  constructor(){}
 *      getFoo() {
 *          return "test";
 *      }
 * }
 *
 * registerFactory(MyFooService, new MyFooService());
 * // or
 * registerFactory({provider: MyFooService, instance: new MyFooService()});
 *
 * @Service()
 * export class OtherService {
 *      constructor(myFooService: MyFooService){
 *          console.log(myFooFactory.getFoo()); /// "test"
 *      }
 * }
 * ```
 *
 */
exports.registerFactory = exports.GlobalProviders.createRegisterFn(interfaces_1.ProviderType.FACTORY);
/**
 * Add a new service in the `ProviderRegistry`. This service will be built when `InjectorService` will be loaded.
 *
 * #### Example
 *
 * ```typescript
 * import {registerService, InjectorService} from "@tsed/common";
 *
 * export default class MyFooService {
 *     constructor(){}
 *     getFoo() {
 *         return "test";
 *     }
 * }
 *
 * registerService({provide: MyFooService});
 * // or
 * registerService(MyFooService);
 *
 * const injector = new InjectorService();
 * injector.load();
 *
 * const myFooService = injector.get<MyFooService>(MyFooService);
 * myFooService.getFoo(); // test
 * ```
 *
 * @param provider Provider configuration.
 */
exports.registerService = exports.GlobalProviders.createRegisterFn(interfaces_1.ProviderType.SERVICE);
/**
 * Register a provider configuration.
 * @param {IProvider<any>} provider
 */
function registerProvider(provider) {
    if (!provider.provide) {
        throw new Error("Provider.provide is required");
    }
    exports.ProviderRegistry.merge(provider.provide, provider);
}
exports.registerProvider = registerProvider;
